File: src/org/redrune/core/task/ScheduledTask.java
package org.redrune.core.task;

import com.google.common.base.Preconditions;
import lombok.Getter;

/**
 * A game-related task that is scheduled to run in the future.
 *
 * @author Graham
 * @author Tyluur <itstyluur@gmail.com>
 * @since 5/26/2017
 */
public abstract class ScheduledTask {
	
	/**
	 * The maximum amount of pulses that can be ran on this task
	 */
	@Getter
	private final int maxPulses;
	
	/**
	 * The delay between executions of the task, in pulses.
	 */
	private int delay;
	
	/**
	 * The number of pulses remaining until the task is next executed.
	 */
	@Getter
	private int pulses;
	
	/**
	 * The amount of times this task has been pulsed
	 */
	@Getter
	private int pulseCount = 0;
	
	/**
	 * A flag indicating if the task is running.
	 */
	private boolean running = true;
	
	/**
	 * Creates a new scheduled task.
	 *
	 * @param delay
	 * 		The delay between executions of the task, in pulses.
	 * @param immediate
	 * 		Indicates whether or not this task should be executed immediately, or after the {@code delay}.
	 * @throws IllegalArgumentException
	 * 		If the delay is less than or equal to zero.
	 */
	public ScheduledTask(int delay, boolean immediate) {
		this(delay, 0, immediate);
	}
	
	/**
	 * Creates a new scheduled task.
	 *
	 * @param delay
	 * 		The delay between executions of the task, in pulses.
	 * @param maxPulses
	 * 		The maximum amount of pulses that this task can go through
	 * @param immediate
	 * 		Indicates whether or not this task should be executed immediately, or after the {@code delay}.
	 * @throws IllegalArgumentException
	 * 		If the delay is less than or equal to zero.
	 */
	public ScheduledTask(int delay, int maxPulses, boolean immediate) {
		setDelay(delay);
		this.pulses = immediate ? 0 : delay;
		this.maxPulses = maxPulses;
	}
	
	/**
	 * Sets the delay.
	 *
	 * @param delay
	 * 		The delay.
	 * @throws IllegalArgumentException
	 * 		If the delay is less than zero.
	 */
	public final void setDelay(int delay) {
		Preconditions.checkArgument(delay >= 0, "Delay cannot be less than 0.");
		this.delay = delay;
	}
	
	/**
	 * Checks if this task is running.
	 *
	 * @return {@code true} if so, {@code false} if not.
	 */
	public final boolean isRunning() {
		return running;
	}
	
	/**
	 * Stops the task.
	 */
	public void stop() {
		running = false;
	}
	
	/**
	 * Pulses this task: updates the delay and calls {@link #execute()} if necessary.
	 */
	final void pulse() {
		if (running && --pulses <= 0) {
			execute();
			pulseCount++;
			pulses = delay;
		}
	}
	
	/**
	 * Executes this task.
	 */
	public abstract void execute();
	
}

File: src/main/kotlin/org/redrune/engine/tick/schedule/ScheduledTask.kt
package org.redrune.engine.tick.schedule

import com.google.common.base.Preconditions

/**
 * A game-related task that is scheduled to run in the future.
 *
 * @author Graham
 * @author Tyluur <itstyluur@icloud.com>
 * @since 5/26/2017
 */
abstract class ScheduledTask @JvmOverloads constructor(delay: Int = 1, goalTicks: Int = 1) {

    /**
     * The maximum amount of ticks that can be ran on this task
     */
    val goalTicks: Int

    /**
     * The delay between executions of the task, in ticks.
     */
    private val delay: Int

    /**
     * The number of ticks remaining until the task is next executed.
     */
    var ticks: Int
        private set

    /**
     * The amount of times this task has been pulsed
     */
    var ticksPassed = 0
        private set

    /**
     * The delayed tick count, this is incremented each time the task is pulsed. The task is pulsed based on the delay
     * set in the constructor.
     */
    var delayedTickCount = 0
        private set
    /**
     * Checks if this task is running.
     *
     * @return `true` if so, `false` if not.
     */
    /**
     * A flag indicating if the task is running.
     */
    var isRunning = true
        private set

    override fun toString(): String {
        return "ScheduledTask{" + "goalTicks=" + goalTicks + ", delay=" + delay + ", ticks=" + ticks + ", ticksPassed=" + ticksPassed + ", running=" + isRunning + '}'
    }

    /**
     * Stops the task.
     */
    protected fun stop() {
        isRunning = false
    }

    /**
     * Pulses this task: updates the delay and calls [Runnable.run] )} if necessary.
     */
    fun pulse() {
        // task wasnt forced to stop
        if (!isRunning) {
            return
        }
        // reduce delay
        ticks--

        // ticks passed increments
        ticksPassed++

        // time until the next one has lapsed
        if (ticks <= 0) {
            // increases the delayed tick count
            delayedTickCount++
            // reset in the case of infinite looping pulse
            ticks = delay
            // runs the task
            run()
        }
    }

    /**
     * Runs the task
     */
    abstract fun run()
    /**
     * Creates a new scheduled task.
     *
     * @param delay
     * The delay between executions of the task, in ticks.
     * @param goalTicks
     * The maximum amount of ticks that this task can go through
     * @throws IllegalArgumentException
     * If the delay is less than or equal to zero.
     */
    /**
     * Creates a new scheduled task.
     *
     * @param delay
     * The delay between executions of the task, in ticks.
     * @throws IllegalArgumentException
     * If the delay is less than or equal to zero.
     */
    /**
     * Constructs a scheduled task with a tick delay
     */
    init {
        Preconditions.checkArgument(delay >= 0, "Delay cannot be negative")
        this.delay = delay
        ticks = delay
        this.goalTicks = goalTicks
    }
}